pkg/controller/deployment/sync.go

```diff
@@ -400,15 +409,185 @@ func (dc *DeploymentController) scale(ctx context.Context, deployment *apps.Depl
+		nameToSizeScalePlan := deploymentutil.PrepareProportionalScalePlan(logger, allActiveRSs, deployment, deploymentReplicasToAdd)
+		nameToSizeScalePlanFullScaleSimulation := nameToSizeScalePlan
+		if deploymentutil.IsDeploymentPodReplacementPolicyEnabled() {
+			// Run a simulation of what would happen during a full scale with empty or TerminationStarted pod replacement policy.
+			// Important only when using the TerminationComplete policy during scaling up to assess partial scaling.
+			if deploymentReplicasToAddFullScaleSimulation > 0 && deploymentutil.HasTerminationCompletePodReplacement(deployment) {
+				nameToSizeScalePlanFullScaleSimulation = deploymentutil.PrepareProportionalScalePlan(logger, allActiveRSs, deployment, deploymentReplicasToAddFullScaleSimulation)
+			}
+		}
+
+		// Update all replica sets
+		for i := range allActiveRSs {
+			rs := allActiveRSs[i]
+
+			newScale := nameToSizeScalePlan[rs.Name]
+			newScaleFullSimulation := nameToSizeScalePlanFullScaleSimulation[rs.Name]
+			// partialScaling is true if there are replica sets that will have to be scaled later due to terminating or surge pods.
+			partialScaling := newScale != newScaleFullSimulation
+
+			// scaleReplicaSet should be called even if there is no replica count update; to update the annotations.
+			// 1. During partial scaling the "deployment.kubernetes.io/replicaset-replicas-before-scale" annotation has to
+			//    be updated to proceed with scaling in the future when terminating pods terminate.
+			// 2. During full scaling the "deployment.kubernetes.io/desired-replicas" annotation has to be updated to mark
+			//    this replica set as fully scaled. When all replica sets are fully scaled, scaling operations will stop.
+			// 3. During full scaling the "deployment.kubernetes.io/max-replicas" has to be updated to prepare the
+			//    replica set for future scaling operation.
+			if _, _, err := dc.scaleReplicaSetV2(ctx, rs, newScale, deployment, partialScaling); err != nil {
+				// Return as soon as we fail, the deployment is requeued
+				return err
+			}
+		}
+	}
+	return nil
+}
+
 func (dc *DeploymentController) scaleReplicaSetAndRecordEvent(ctx context.Context, rs *apps.ReplicaSet, newScale int32, deployment *apps.Deployment) (bool, *apps.ReplicaSet, error) {
 	// No need to scale
 	if *(rs.Spec.Replicas) == newScale {
 		return false, rs, nil
 	}
-	scaled, newRS, err := dc.scaleReplicaSet(ctx, rs, newScale, deployment)
-	return scaled, newRS, err
+	if deploymentutil.IsDeploymentPodReplacementPolicyEnabled() {
+		scaled, newRS, err := dc.scaleReplicaSetV2(ctx, rs, newScale, deployment, false)
+		return scaled, newRS, err
+	} else {
+		scaled, newRS, err := dc.scaleReplicaSet(ctx, rs, newScale, deployment)
+		return scaled, newRS, err
+	}
 }
 
+func (dc *DeploymentController) scaleReplicaSetV2(ctx context.Context, rs *apps.ReplicaSet, newScale int32, deployment *apps.Deployment, partialScaling bool) (bool, *apps.ReplicaSet, error) {
```

Comment:
This `scaleReplicaSetV2` is worrisome. This will kick in every time we have the FG enabled, so we have two possible paths for the same functionality (the case when FG is on and no podReplacementPolicy is set, or when the FG is off). The only difference I see is two function invocations: 
1. `ComputeReplicaSetScaleAnnotationsV2` vs `ReplicasAnnotationsNeedUpdate`
2. `SetReplicaSetScaleAnnotationsV2` vs `SetReplicasAnnotations`

I think it'd be easier to digest if we modify the current path such that it looks more like the new proposed functions. Then the additional changes would be gated through the new FG, only.

Reply:
Personally, I would prefer to unify the annotation utility functions as well, but this would involve touching the existing code paths. Thus increasing a chance of a regression.

I would like to hear @liggitt opinion on the annotations before proceeding any further.